backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/settings/ModularContractSettingsDAO.java

```diff
@@ -142,4 +157,212 @@ public boolean isSettingsExist(final @NonNull ModularContractSettingsQuery query
+	{
+		@Override
+		public Collection<SettingsLookupKey> computeCachingKeys(@NonNull final TableRecordReference recordRef)
+		{
+			if (I_C_Flatrate_Conditions.Table_Name.equals(recordRef.getTableName()))
+			{
+				logger.debug("ComputeCachingKeys called for a ({},{}) that wasn't cached so far!",
+							 recordRef.getRecord_ID(),
+							 recordRef.getTableName());
+
+				return ImmutableSet.of();
+			}
+			else if (I_C_Flatrate_Term.Table_Name.equals(recordRef.getTableName()))
+			{
+				final FlatrateTermId contractId = recordRef.getIdAssumingTableName(I_C_Flatrate_Term.Table_Name, FlatrateTermId::ofRepoId);
+
+				logger.debug("ComputeCachingKeys called for ({},{})!",
+							 recordRef.getRecord_ID(),
+							 recordRef.getTableName());
+
+				return ImmutableSet.of(SettingsLookupKey.of(contractId));
+			}
+
+			throw new AdempiereException("Unexpected table name=" + recordRef.getTableName());
+		}
+
+		@Override
+		public boolean isResetAll(@NonNull final TableRecordReference recordRef)
+		{
+			if (I_C_Flatrate_Term.Table_Name.equals(recordRef.getTableName()))
+			{
+				return false;
+			}
+			else if (I_C_Flatrate_Conditions.Table_Name.equals(recordRef.getTableName())
+					&& resetAllForFlatrateConditions(recordRef.getIdAssumingTableName(I_C_Flatrate_Conditions.Table_Name, ConditionsId::ofRepoId)))
+			{
+				return true;
+			}
+
+			//reset all the cache if the object was deleted, and we can't get to a SettingsLookupKey
+			final Object recordRefModel = recordRef.getModel(Object.class);
+			return recordRefModel == null;
+		}
+
+		private boolean resetAllForFlatrateConditions(@NonNull final ConditionsId conditionsId)
+		{
+			return cacheKey2SettingsId.containsKey(SettingsLookupKey.of(conditionsId));
+		}
+	}
+
+	@Value
+	private static class SettingsLookupKey
+	{
+		@Nullable
+		FlatrateTermId contractId;
+		@Nullable
+		ConditionsId conditionsId;
+
+		@NonNull
+		public static SettingsLookupKey of(@NonNull final ConditionsId conditionsId)
+		{
+			return new SettingsLookupKey(null, conditionsId);
+		}
+
+		@NonNull
+		public static SettingsLookupKey of(@NonNull final FlatrateTermId contractId)
+		{
+			return new SettingsLookupKey(contractId, null);
+		}
+
+		private SettingsLookupKey(
+				@Nullable final FlatrateTermId contractId,
+				@Nullable final ConditionsId conditionsId)
+		{
+			if (conditionsId == null && contractId == null)
+			{
+				throw new AdempiereException("conditionsId && contractId cannot be both null!");
+			}
+
+			if (conditionsId != null && contractId != null)
+			{
+				throw new AdempiereException("conditionsId && contractId cannot be both set!");
+			}
+
+			this.contractId = contractId;
+			this.conditionsId = conditionsId;
+		}
+	}
+
+	@Value
+	private static class CachedSettingsId
+	{
+		private static CachedSettingsId ofNullable(@Nullable final ModularContractSettingsId settingsId)
+		{
+			return new CachedSettingsId(settingsId);
+		}
+
+		@Nullable
+		ModularContractSettingsId settingsId;
+	}
```

Comment:
using this wrapper class, I wanted to avoid querying the database twice for the same termId/conditionsId in case there is no settings linked to them. 

I have to say, it's not ideal, so... I'm open to suggestions :)